### ISSUE TITLE
vaapi fallback if gpu too slow?

### ISSUE THREAD

**Issue by fefe17** [mpv-player/mpv#6301]
I'm using mpv from git on an Acer Swift 1 under Linux with X11.
This is a $700 notebook with a full HD display and a Gemini Lake 4-core newer-gen Atom (Pentium N5000). It's a very nice hardware with a battery that can do 10+ hours of non-stop video playback.
But not on Linux with mpv, because mpv uses "gpu" by default, which produces amazingly choppy video (multiply skipped frames per sec) and then tells my my driver is probably broken.

So I went to the Wiki and it said "gpu" is much better than vaapi and also uses vaapi and nobody should be using vaapi directly as the quality is better with gpu.

I respectfully disagree. This laptop has a power usage of 5W with full HD video playback under Linux with mpv -vo vaapi. It can play back full HD video just fine with -vo xv, but not with -vo gpu.

I am using the latest Mesa + Intel drivers.

My suggestion: If mpv notices that playback with -vo gpu is choppy and drops frames, change the error message to suggest trying -vo vaapi, too.

And change the wiki to mention that using vaapi is potentially much more battery-friendly than -vo gpu or even -vo gpu-hq.

**Comment by CounterPillow** [comment]
>And change the wiki to mention that using vaapi is potentially much more battery-friendly than -vo gpu or even -vo gpu-hq.

Cue laptop people all switching to the vaapi VO because the wiki told them it's "battery friendly" and then opening issues about how HDR tonemapping or debanding doesn't work.

I wrote that article, and the reason why it tells people very strongly to use `gpu` is because if people do not know what VO they should be using, they should most likely be using `gpu`. As the official mpv docs state on the vaapi VO: "This driver is for compatibility with crappy systems." and "This is low quality, and has issues with OSD." I've seen enough people on IRC asking for help who were using the vaapi VO for no reason, hence why the article was written in the first place.

If the `gpu` VO doesn't work smoothly on your system, it means you're on a crappy system. Battery savings shouldn't be a reason to recommend a VO where all the mpv rendering is bypassed, since users usually don't seem to realise what this means and think setting `profile=gpu-hq` will still give them better video rendering using such VOs.

**Comment by fefe17** [comment]
Respectfully, sir, it's not your place to tell me if my hardware is shitty or not.

Also: THE BALLS on you to tell me that -vo gpu produces better quality when it drops half the frames.

I bought this laptop because it can play back 10 hours of video on battery.
And it can. So it is by definition not crappy. It does exactly what I bought it for.

Just not using mpv.

So you have a choice now. You keep lying in the wiki, and people will then switch from mpv on Linux to vlc or Windows, both of which can do the playback easily and without trying to "educate" me. Or you can be truthful in your wiki entry and people keep using mpv.

How about you leave out the advocacy and poltics and tell the facts.
-vo gpu is the default because most laptops (that are also built for 3d graphics) have enough horsepower and it has these advantages: *list*
However, it achieves those advantages by potentially using more battery power (find out by running powertop on Linux) and/or creating more load on the GPU. For low-end systems optimized for video playback instead of 3d graphics, that load may be too much for the system and mpv will start dropping frames.

**Comment by haasn** [comment]
Sounds like this is confused between a bug report and a feature request, I'll lean towards the former, since it sounds very much a case of "my broken system can't do opengl please fix". No way to debug. Closing.

**Comment by CounterPillow** [comment]
>Just not using mpv.

feel free to do that

**Comment by CounterPillow** [comment]
If your hardware cannot do a bilinear scale of a texture the size of the luma and chroma planes then it is objectively crappy hardware. Alternatively he did something stupid like use llvmpipe as an opengl implementation, though we cannot actually know that since he ignored the issue template. I'd be surprised if vaapi video output on up to date Intel GPUs even still had dedicated hardware and didn't just use the GPU silicon for scaling.

>he made a constructive suggestion

No, he was acting like an entitled prick. I'm guessing you're one of his friends who he talked into brigading here.

**Comment by nero** [comment]
@CounterPillow this issue was mentioned on a tech blog popular in germany: [http://blog.fefe.de/?ts=a51b4ed1](http://blog.fefe.de/?ts=a51b4ed1). Thats their fanclub showing up here.

**Comment by rdggithub** [comment]
> I'm guessing you're one of his friends who he talked into brigading here.

> Thats their fanclub showing up here.

I read through this issue and found the final answer connoting something he didn't write. And I wondered where the anger came from. That's all, mates.

He explained it to me ... "issue template", "acting like..." ... that's fine for me. No reason to be rude.

**Comment by CounterPillow** [comment]
@nero ayy lmao so it is that fefe, that explains some things.

**Comment by Akemi** [comment]
we as a project _respectfully_ disagree with you regarding vaapi and recommending it for any kind of usage besides the ones already mentioned in the manual. we can't recommend a vo, or any part of our code, that is not actively maintained and is only there because it's not in the way and for compatibility. one could even say it's quasi deprecated.

there should be a reason why vaapi works better for you than gpu. just claiming it's slower, etc without really backing up your claim will just lead to a dead end. since you ignored our issue template we will assume that you are not interested in resolving the underlying issue and just wanted to make a suggestion, the we as a project _respectfully_ decline.

if you are happy with the result you get with vaapi we _happily_ gratulate you for finding your personal sweet spot and thank you for using our software.

on a side note @CounterPillow is not an _official representative_ of our project, but rather a resident troll that is very direct in his speech.

again thank you for using our software and we hope to be of assistance again.

**Comment by jeeb** [comment]
**tl;dr**
1. No, `vo=vaapi` cannot be automagically recommended because it requires VAAPI capable hardware to be found, as well as has known draw-backs due to just feeding the video to your GPU's fixed function rendering pipe line that is 100% out of control with regards to pretty much anything. If it handles your use cases well enough, great. Unfortunately that is not always the case and I don't think you even need to go for the fancy features for that (even handling color spaces like BT.2020 or 10bit decoded video can be problematic depending on the driver and/or hardware).
2. `vo=gpu` is not for gaming hardware only! My 2014 Adreno Android device is a living example of that, together with quite a few laptops/tablets I have from 2008 to 2015. But it does use the GPU part of your GPU, as opposed to the fixed function pipe lines. If there's a problem in the driver or the hardware with the back-end API you are using (by default OpenGL for *nix, D3D11 for Windows), then it will not be fast enough.
2. Unfortunately the code cannot yet figure out **why** seemingly A/V synchronization got broken. This means that it cannot figure out what you should do, either. We leave this to humanIf you're utilizing various things most optimally, or getting slowed down by decoding on CPU, or whatever else there could be (and trust me, it's software - there can be a **lot** of reasons for things seemingly not going perfectly). The message mentions a lot of possible reasons and is IMHO quite respectful. If you had questions about **why** you were getting this stuff, why did you not come to ask about it and providing information/logs about the setup and configuration etc?
   ```c
   "Audio/Video desynchronisation detected! Possible reasons include too slow\n"
   "hardware, temporary CPU spikes, broken drivers, and broken files. Audio\n"
   "position will not match to the video (see A-V status field).\n"
   ```
3. Follow the issue template if you want support. Post logs and make it simpler for people to understand what your issue is (at this point we don't even know if you had hwdec enabled with `vo=gpu` while you were comparing it with `vo=vaapi` which most likely had!). Also make an effort to understand _why_ things are like they are. For more free-form communication there's the IRC channel.

**Original crap I kept writing**

Additionally, if you had posted a log and/or appeared on one of the active support areas (such as the IRC channel), someone could have at least looked if you were actually using hwdec in both cases. Or if you were using hwdec in the most optimal way with any given renderer.

I also have to mention that `vo=gpu` is not a "gamer GPU" thing at all, as proved by the fact that I can watch my 1080p cartoons on a 2014 Android Adreno device that cost <400eur back then just fine by letting MediaCodec do the decoding, and `vo=gpu` do the rendering (although swdec does generally work as well, albeit not as great for battery life for obvious reasons on a low-power ARM device). It also works on my Intel tablet as well on both Windows and Linux. `vo=gpu` **can** be fancy, but the defaults are far from it, and [additional options](https://mpv.io/manual/master/#options-gpu-dumb-mode) are mentioned in the manual.

Now, if the end result was that you had tested with the latest code base, required hwdec and were using it most optimally with `vo=gpu` and it still wasn't cutting it, then unfortunately the graphics driver or the graphics hardware just couldn't handle it(*). I know there are some devices where this happens in general (my 2008 eeepc was one of them back in the day), but generally speaking they are - or hopefully should be rare. I would guess such hardware is still more prevalent in the ARM space where in hardware budgeting nobody expects you to do custom video rendering, but instead to utilize your graphics chip's overlay features (which you have no control over if they handle the colorspaces etc correctly or not - same as with the VAAPI video rendering API).

(* As a funny fact, various devices that have bad OpenGL support but support Direct3D 11 on Windows most likely have quite a bit better performance with dxva2/d3d11va hwdec and the (default) Direct3D 11 back-end for `vo=gpu`. So the hardware might not be bad, just the drivers.)

And here we come to some of the basic tenets of mpv:
1. Try to stay in control of the playback chain
2. Try to provide a consistent experience to the user where guessing if a vendor implemented X or Y correctly should not be a thing (compare this to DirectShow playback on Windows for example with the standard MS filters).

This is why:
1. `vo=gpu` is recommended. It requires some amount of GPU power (not gamer level power as mentioned!), but enables one to get a relatively consistent experience on most systems. Still, unless things actively get in the way (block some sort of improvement) we do have a habit of [keeping alternatives](https://mpv.io/manual/master/#video-output-drivers) that give up control for (possible) performance (x11/xv/vaapi are some of these for *nix, direct3d for Windows).
2. `hwdec` is specifically disabled by default. While hwdec is better than it was before (circa 2010), using software decoding is still the only way to have a consistent. _Yet we still try to optimize the hwdec paths in `vo=gpu` as there are real-life use cases where having (proper) decoding available through hardware is preferable._ In the manual we even have [a list](https://mpv.io/manual/master/#options-hwdec) of how much damage different hwdec modes can cause (scroll down a bit from that anchor point).

**Comment by haasn** [comment]
By the way, for somebody who seemingly is so concerned with the public appearance of FOSS projects, [it doesn't exactly send a positive image to lie about others online](http://blog.fefe.de/?ts=a51b4ed1).

**Comment by lachs0r** [comment]
@fefe17: Just to make it absolutely clear (and I am repeating what has been mentioned here already), we have an issue template for a reason. Ignoring it (which is reason enough to close an issue as invalid) and then acting like you’re entitled to every last minute of our time while not providing the information that would have been useful in diagnosing the cause of your playback issues while also trying to publically shame the project for the behavior of a single person who certainly does not represent it (note his Contributor badge: it doesn’t say “Member”) does not put you in a good bargaining position. It should come as no surprise that the issue gets locked when someone’s “fan club” gets involved and threatens to derail the thread.

Just as you get tired from having to go through registration procedures and so on just to report a bug, we are getting tired of people not even putting in the effort of reading the issue template, preferring to waste everyone’s time by starting useless debates instead. We *are* happy to fix issues users are having with mpv, *if it is within our power to do so*. We can’t do shit without even knowing what the problem is (believe me, it’s not always as simple as it seems). You have to help us figure it out.

I hope you can at least tell that some of us make an honest effort to be reasonable and help you resolve your issue despite all that. You are welcome to try again without ignoring the template, and without posting immature rants on your blog because you feel so important.
